libr/util/r_pkcs7.c

```diff
@@ -0,0 +1,372 @@
+#include <stdlib.h>
+#include <string.h>
+#include <r_util.h>
+
+bool r_pkcs7_parse_certificaterevocationlists (RPKCS7CertificateRevocationLists *crls, RASN1Object *object) {
+	ut32 i;
+	if (!crls && !object) {
+		return false;
+	}
+	if (object->list.length > 0) {
+		crls->elements = (RX509CertificateRevocationList **) calloc (object->list.length, sizeof (RX509CertificateRevocationList*));
+		if (!crls->elements) {
+			return false;
+		}
+		crls->length = object->list.length;
+		for (i = 0; i < crls->length; ++i) {
+			crls->elements[i] = r_x509_parse_crl (object->list.objects[i]);
+		}
+	}
+	return true;
+}
+
+void r_pkcs7_free_certificaterevocationlists (RPKCS7CertificateRevocationLists *crls) {
+	ut32 i;
+	if (crls) {
+		for (i = 0; i < crls->length; ++i) {
+			r_x509_free_crl (crls->elements[i]);
+			crls->elements[i] = NULL;
+		}
+		free (crls->elements);
+		crls->elements = NULL;
+		// Used internally pkcs #7, so it should't free crls.
+	}
+}
+
+bool r_pkcs7_parse_extendedcertificatesandcertificates (RPKCS7ExtendedCertificatesAndCertificates *ecac, RASN1Object *object) {
+	ut32 i;
+	if (!ecac && !object) {
+		return false;
+	}
+	if (object->list.length > 0) {
+		ecac->elements = (RX509Certificate **) calloc (object->list.length, sizeof (RX509Certificate*));

```

Comment:
The length value is calculated from the DER structure; i just parse it and fill things into my structures. is true that can overflow, but i've added checks here: https://github.com/wargio/radare2/blob/b397776d267d5f2fec2bfd94d9f1472b71d504e1/libr/util/r_asn1.c#L324

It can't overflow (from my point of view).